pennylane/ops/functions/iterative_qpe.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the function which generates the iterative QPE subroutine.
+"""
+
+import pennylane as qml
+import numpy as np
+
+
+def iterative_qpe(base, estimation_wire, iters):
+    r"""Function that is in charge of adding to the queue the necessary gates to implement iterative QPE.
+    It returns a list of the mid-circuit measurements performed.
+
+    Args:
+      base (Operator): the phase estimation unitary, specified as an :class:`~.Operator`
+      estimation_wire (Union[Wire, or int]): the wire to be used for the estimation.
+      iters (int): the number of measurements to be performed.
+
+    Returns:
+      list[MidMeasureMP]: the list of measurements performed.
+
+    .. seealso:: :class:`~.QuantumPhaseEstimation`
+
+    **Example**
```

Comment:
Maybe an advanced example with multiple shots to show users what they expect to see?

Reply:
Curious to see what the advanced example looks like - also as validation that IQPE works as we expect.

If a more advanced example is too long, it could go into a usage details section. Otherwise, best left to a demo.

Reply:
The advanced example is not too pretty right now. We would have to wait until we can have generalized statistics on the measurements. However, if needed now, the post-processing would be like this:

```python
dev = qml.device("default.qubit", shots=1000)

@qml.qnode(dev)
def circuit_iterative():
    # Initial state
    qml.PauliX(wires=[0])

    # Iterative QPE
    measurements = qml.iterative_qpe(qml.RZ(phi, wires=[0]), estimation_wire=[1], iters=3)

    return [qml.sample(op=meas) for meas in measurements]

# This is the part that will change hopefully. <--------
sample_list = np.array(circuit_iterative())
sample_list = sample_list.T
output = qml.probs().process_samples(np.array(sample_list), wire_order=[0, 1, 2])

```